src/coreclr/jit/lsrabuild.cpp

```diff
@@ -3846,9 +3846,23 @@ int LinearScan::BuildDelayFreeUses(GenTree*         node,
             return 0;
         }
     }
+
```

Comment:
@kunalspathak : As suggested offline, added some additional checks in the delay slot logic. If the register type of the register does not match that of the delay slot register then do not add the delay free use. This fixes where we call `BuildDelayFreeUses()` for `op2` which is a integer value.

Instead, we could do these checks in `BuildHWIntrinsic()` to not call `BuildDelayFreeUses()`, but then `BuildDelayFreeUses()` would still need the same checks turned into asserts. So it seemed simpler to do inside.

This will work nicely with `BuildHWIntrinsic()` rewrite PR too.

Reply:
Ideally this should be assert and the callers should not be calling delay free uses for different register type. Did you turned it into an assert and see how many places are hit?

Reply:
Difficult to test with an assert because 1) all the hwintrinsic tests will fail at the first assert 2) many of these issues will be due during `ConstantExpected` APIs, and for a lot of those we are only testing using hardcoded constants, meaning the assert will never be hit.

Scanning the SVE API, all the methods we have that are `RMW` and have a `ConstantExpected` are:
```
AddRotateComplex
DotProductBySelectedScalar
ExtractVector
FusedMultiplyAddBySelectedScalar
FusedMultiplySubtractBySelectedScalar
MultiplyAddRotateComplex
MultiplyAddRotateComplexBySelectedScalar
MultiplyBySelectedScalar
SaturatingDecrementBy16BitElementCount
SaturatingDecrementBy32BitElementCount
SaturatingDecrementBy64BitElementCount
SaturatingDecrementBy8BitElementCount
SaturatingDecrementByActiveElementCount
SaturatingIncrementBy16BitElementCount
SaturatingIncrementBy32BitElementCount
SaturatingIncrementBy64BitElementCount
SaturatingIncrementBy8BitElementCount
SaturatingIncrementByActiveElementCount
ShiftRightArithmeticForDivide
TrigonometricMultiplyAddCoefficient
```

Then there are AdvSimd ones. Then there are possibly others.

Reply:
This is much easier to fix properly in the [lsra rewrite PR](#107459), as the check can just be added into the main for loop in `BuildHWIntrinsic()`

```
        else if (delayFreeOp != nullptr && TheExtraChecksFromThisPR.....)
        {
            srcCount += BuildDelayFreeUses(operand, delayFreeOp, candidates);
        }
        else
        {
            srcCount += BuildOperandUses(operand, candidates);
        }
```